blocks/bricks/recurrent.py

```diff
@@ -630,3 +631,231 @@ def apply(self, *args, **kwargs):
+        for transition in self.transitions:
+            transition.push_allocation_config()
+
+        for d, fork in enumerate(self.forks):
+            fork.input_dim = self.transitions[d].get_dim(self.states_name)
+            fork.output_dims = self.transitions[d+1].get_dims(
+                fork.output_names)
+        for d, distribute in enumerate(self.distributes):
+            distribute.input_dims = self.transitions[0].get_dims(
+                self.normal_inputs(0))
+            distribute.output_dims = self.transitions[d+1].get_dims(
+                self.normal_inputs(d+1))
+
+    def do_apply(self, *args, **kwargs):
+        """Apply the stack of transitions.
+
+        This is the undecorated implementation of the apply method.
+        It is separated from the decorated apply method in order to allow
+        usage of different docrations (wrappers) to be used.
+
+
+        Parameters
+        ----------
+        iterate : bool
+            If ``True`` iteration is made. By default ``True``.
+        reverse : bool
+            If ``True``, the sequences are processed in backward
+            direction. ``False`` by default.
+        return_initial_states : bool
+            If ``True``, initial states are included in the returned
+            state tensors. ``False`` by default.
+        The names appearing in the sequences, states, contexts parameters
+            of the apply method of each of the transitions in
+            self.transitions. Each such name is suffixed with a layer
+            number.
+
+        Returns
+        -------
+        The outputs of all transitions.
+        All return values are of type :class:`~tensor.TensorVariable`.
+
+        """
+        results = []
+        last_states = None
+        layer0_kwargs = None
+        for d, transition in enumerate(self.transitions):
+            sequences_names = transition.apply.sequences
+            if d == 0:
+                layer0_kwargs = dict(
+                    (s, kwargs.get(s)) for s in sequences_names)
+                layer_kwargs = dict(layer0_kwargs)
+            else:
+                inputs = self.forks[d-1].apply(last_states, as_list=True)
+                if self.distributes:
+                    distributes = self.distributes[d-1].apply(as_list=True,
+                                                              **layer0_kwargs)
+                    for i in range(len(inputs)):
+                        inputs[i] += distributes[i]
+                layer_kwargs = dict(zip(self.normal_inputs(d), inputs))
+                if "mask" in sequences_names:
+                    layer_kwargs["mask"] = kwargs.get("mask")
+
+            for t in ["states", "contexts", "outputs"]:
+                suffix = '_' + str(d)
+                for s in getattr(transition.apply, t):
+                    layer_kwargs[s] = kwargs.get(s + suffix)
+
+            for k in ['iterate', 'reverse', 'return_initial_states']:
```

Comment:
Yes it is bad but this method (do_apply) in its fast_apply mode is mimicking the behaviour of recurrent_apply so it should be aware of its argumenrs somehow (any other idea will be welcomed.)